server/task.go

```diff
@@ -36,6 +36,26 @@ func isTaskStatusTransitionAllowed(fromStatus, toStatus api.TaskStatus) bool {
 	return false
 }
 
+func (s *Server) canUpdateTaskStatement(ctx context.Context, task *api.Task) *echo.HTTPError {
+	// Allow frontend to change the SQL statement of
+	// a PendingApproval task which hasn't started yet
+	// a Failed task which can be retried
+	// a Pending task which can't be scheduled because of failed task checks
+	if task.Status != api.TaskPendingApproval && task.Status != api.TaskFailed && task.Status != api.TaskPending {
+		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("can not update task in %v state", task.Status))
+	}
+	if task.Status == api.TaskPending {
+		schedule, err := s.TaskScheduler.canScheduleTask(ctx, task)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, "failed to check whether the task can be scheduled").SetInternal(err)
+		}
+		if !schedule {
+			return echo.NewHTTPError(http.StatusBadRequest, "can not update the PENDING task because it can be running at any time")
+		}
+	}
```

Comment:
```suggestion
	if task.Status == api.TaskPending {
		canSchedule, err := s.TaskScheduler.canScheduleTask(ctx, task)
		if err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, "failed to check whether the task can be scheduled").SetInternal(err)
		}
		if canSchedule {
			return echo.NewHTTPError(http.StatusBadRequest, "can not update the PENDING task because it can be running at any time")
		}
	}
```
IMO, if we can schedule a pending task, we should not let the user change the SQL statement.